File: HW/HW5/D.cpp
#include <iostream>

using namespace std;

#define N 100005
#define SQRT 100

int A[N];
long long dp[N][SQRT];
int n;

long long solve(int s, int k) {
    if (k >= SQRT) { // Calculate sum iteratively
        long long sum = 0;
        for (int i = s; i < n; i += k) sum += A[i];
        return sum;
    } else { // Just return preprocessed value
        return dp[s][k];
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);    

    cin >> n;
    for (int i = 0; i < n; i++) cin >> A[i];

    // Preprocess
    for (int s = n - 1; s >= 0; s--) {
        for (int k = 1; k < SQRT; k++) {
            if (s + k < n) dp[s][k] = dp[s + k][k] + A[s];
            else dp[s][k] = A[s];
        }
    }

    int q; cin >> q;
    for (int i = 0; i < q; i++) {
        int s, k; cin >> s >> k; s--;
        cout << solve(s, k) << '\n';
    }
    return 0;
}

File: HW/HW5/D.py
SQRT = 320
N = 100005
dp = [[0] * SQRT for i in range(N)]

n = int(input())
A = list(map(int, input().split()))

def solve(s, k):
    if k * k > n:
        sum = 0
        for i in range(s, n, k):
            sum += A[i]
        return sum
    else:
        return dp[s][k]

for s in range(n - 1, -1, -1):
    k = 1
    while k * k <= n:
        if s + k < n:
            dp[s][k] = dp[s + k][k] + A[s]
        else:
            dp[s][k] = A[s]
        k += 1

q = int(input())
for i in range(q):
    s, k = map(int, input().split())
    print(solve(s-1, k))